src/kem/kyber/ntt.rs

```diff
@@ -106,72 +73,101 @@ pub(in crate::kem::kyber) fn ntt_binomially_sampled_ring_element(
+    ntt_at_layer!(2, zeta_i, re, 3328);
+    ntt_at_layer!(1, zeta_i, re, 3328);
 
     re.coefficients = re.coefficients.map(barrett_reduce);
 
     re
 }
 
+/// Compute the inverse NTT. The coefficients of the output ring element are in
+/// the Montgomery domain.
 #[inline(always)]
 pub(crate) fn invert_ntt_montgomery<const K: usize>(
     mut re: PolynomialRingElement,
 ) -> PolynomialRingElement {
     // We only ever call this function after matrix/vector multiplication
-    let initial_coefficient_bound = (K as i32) * FIELD_MODULUS;
     hax_lib::debug_assert!(re
         .coefficients
         .into_iter()
-        .all(|coefficient| coefficient.abs() < initial_coefficient_bound));
+        .all(|coefficient| coefficient.abs() < (K as i32) * 3329));
 
-    let mut zetas_index = COEFFICIENTS_IN_RING_ELEMENT / 2;
+    let mut zeta_i = COEFFICIENTS_IN_RING_ELEMENT / 2;
 
-    (re, zetas_index) = invert_ntt_at_layer(1, zetas_index, re, initial_coefficient_bound);
-    (re, zetas_index) = invert_ntt_at_layer(2, zetas_index, re, initial_coefficient_bound);
-    (re, zetas_index) = invert_ntt_at_layer(3, zetas_index, re, initial_coefficient_bound);
-    (re, zetas_index) = invert_ntt_at_layer(4, zetas_index, re, initial_coefficient_bound);
-    (re, zetas_index) = invert_ntt_at_layer(5, zetas_index, re, initial_coefficient_bound);
-    (re, zetas_index) = invert_ntt_at_layer(6, zetas_index, re, initial_coefficient_bound);
-    (re, _) = invert_ntt_at_layer(7, zetas_index, re, initial_coefficient_bound);
+    macro_rules! invert_ntt_at_layer {
+        ($layer:literal) => {
+            let step = 1 << $layer;
+
+            for round in 0..(128 / step) {
+                zeta_i -= 1;
+
+                let offset = round * step * 2;
+
+                for j in offset..offset + step {
+                    let a_minus_b = re.coefficients[j + step] - re.coefficients[j];
+
+                    // Instead of dividing by 2 here, we just divide by
+                    // 2^7 in one go in the end.
+                    re.coefficients[j] = re.coefficients[j] + re.coefficients[j + step];
+                    re.coefficients[j + step] =
+                        montgomery_reduce(a_minus_b * ZETAS_TIMES_MONTGOMERY_R[zeta_i]);
+                }
+            }
+        };
+    }
 
+    invert_ntt_at_layer!(1);
+    invert_ntt_at_layer!(2);
+    invert_ntt_at_layer!(3);
+    invert_ntt_at_layer!(4);
+    invert_ntt_at_layer!(5);
+    invert_ntt_at_layer!(6);
+    invert_ntt_at_layer!(7);
+
+    hax_lib::debug_assert!(
+        re.coefficients[0].abs() < 128 * (K as i32) * 3329
+            && re.coefficients[1].abs() < 128 * (K as i32) * 3329
+    );
     hax_lib::debug_assert!(re
         .coefficients
         .into_iter()
         .enumerate()
         .skip(2)
-        .all(|(i, coefficient)| coefficient.abs() < (128 / (1 << i.ilog2())) * FIELD_MODULUS));
+        .all(|(i, coefficient)| coefficient.abs() < (128 / (1 << i.ilog2())) * 3329));
```

Comment:
Why don't you use the `FIELD_MODULUS`?
Let's use constants whenever possible and add comments when not.

Reply:
Ah, I made that change, but it got lost with the revert.